### DISCUSSION TITLE
Prop drilling을 해결하기 위한 방법에 대해 설명해 주세요

### DISCUSSION THREAD

**Discussion by kamwoo** [woowacourse-study/2022-woowahan-bansanghwe#115] [React]
리액트는 단방향 데이터 흐름이기 때문에 상위 컴포넌트에서 자식 컴포넌트로 계속해서 props를 내려주게 되고 depth가 깊어지게 되면 props를 오로지 하위 컴포넌트로 전달하는 용도로만 쓰이는 컴포넌트들이 많이 생기게 된다. 이러한 현상을 props drilling이라고 한다. 

prop drilling이 보통 3~5개 컴포넌트를 거치는 정도이면, 괜찮을 수 있으나, 10개 이상을 거친다면 props의 갯수가 많아지고, 하위 컴포넌트에서 상위 컴포넌트로 이벤트 콜백을 넘겨주기 위해서 의미없는 많은 단계를 거쳐야하기 때문에 비효율적이고 유지보수하기 어렵기때문이다. 

props drilling을 해결하는 방법

전역 상태 관리

하위의 컴포넌트 전반에 걸쳐서 공유할 수 있는 상태를 말한다. 리액트에서 제공하는 도구로 Context API가 있고, 다양한 전역 상태 라이브러리가 존재한다.

먼저 Context API는 다음과 같이 사용할 수 있다.

```tsx
// messageContext.jsx
import React, { createContext, useEffect, useState } from 'react';

export const messageContext = createContext();

const MessageContextProvider = ({ children }) => {
  const [message, setMessage] = useState('');

  return (
    <messageContext.Provider
      value={{
        message,
        setMessage,
      }}>
      {children}
    </messageContext.Provider>
  );
};

export default MessageContextProvider;
```

```tsx
// MessageComponent.jsx
import { useContext } from 'react';
import { messageContext } from '../../context/MessageContext';

const MessageComponent = () => {
  const messageState = useContext(messageContext);

  return (
    <div>
      {messageState?.message}
    </div>
  );
};

export default MessageComponent;
```

주의점

provider의 상위 컴포넌트에서 리랜더링이 발생하면 그 하위에서 구독하는 있는 컴포넌트 모두 다시 렌더링 된다.

단점

1. provider hell
2. 하위 컴포넌트 리렌더링
3. 재사용성이 어려워짐

**Comment by airman5573** [comment]
바닐라 자바스크립트로 웹앱을 만드는것과 리액트로 웹앱을 만드는것의 차이점은?

**Comment by airman5573** [comment]
서버와 클라이언트의 상태의 sync를 맞추려면 어떻게 해야 할까요?

**Comment by airman5573** [comment]
리액트를 개발하면서 어려웠던 점은 무엇인가요?
